File: windows/VideoSynopsis/VideoSynopsisLibrary/Detector.cpp
#pragma warning(push, 0)
#include "Detector.h"
#include <opencv2/imgproc.hpp>
#include <chrono>
//#include <opencv2/cudabgsegm.hpp>
//#include <opencv2/cudaarithm.hpp>
//#include <opencv2/cudafilters.hpp>
#pragma warning(pop)


Detector::Detector()
{
	//bgSubtractor = cv::cuda::createBackgroundSubtractorMOG2();
	bgSubtractor = cv::createBackgroundSubtractorMOG2();
}

void Detector::processFrame(const cv::UMat& inputFrame)
{
	auto start = std::chrono::high_resolution_clock::now();

	//cv::Mat NinputFrame = inputFrame.getMat(cv::ACCESS_READ);

	cv::UMat rawFg;
	bgSubtractor->apply(inputFrame, rawFg);

	cv::UMat threshold;
	cv::threshold(rawFg, threshold, 127, 255, cv::ThresholdTypes::THRESH_BINARY);

	cv::UMat blur;
	cv::medianBlur(threshold, blur, 5);

	auto structEleme = cv::getStructuringElement(cv::MorphShapes::MORPH_RECT, cv::Size(5, 5));
	auto structEleme2 = cv::getStructuringElement(cv::MorphShapes::MORPH_RECT, cv::Size(9, 9));

	cv::UMat morph1;
	cv::UMat morph2;

	cv::morphologyEx(blur, morph1, cv::MorphTypes::MORPH_CLOSE, structEleme);
	cv::morphologyEx(morph1, morph2, cv::MorphTypes::MORPH_OPEN, structEleme2);

	std::vector<std::vector<cv::Point> > contours;
	std::vector<cv::Vec4i> hierarchy;
	cv::findContours(morph2, contours, hierarchy, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);


	cv::Mat filterMask = cv::Mat::zeros(inputFrame.size(), CV_8UC1);
	for (int i = 0; i < contours.size(); i++)
	{
		double area = cv::contourArea(contours[i]);
		if (area > 20) {
			cv::drawContours(filterMask, contours, i, cv::Scalar(255), -1);
		}
	}

	cv::Mat fiiltered;

	auto structEleme3 = cv::getStructuringElement(cv::MorphShapes::MORPH_ELLIPSE, cv::Size(29, 29));
	cv::morphologyEx(filterMask, fiiltered, cv::MorphTypes::MORPH_DILATE, structEleme3);

	cv::Mat filterMask2 = cv::Mat::zeros(inputFrame.size(), CV_8UC3);
	cv::findContours(fiiltered, contours, hierarchy, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

	for (int i = 0; i < contours.size(); i++)
	{
		double area = cv::contourArea(contours[i]);
		if (area > 20)
		{
			cv::rectangle(filterMask2, cv::boundingRect(contours[i]), cv::Scalar(255,255,255), -1);
		}
	}


	cv::bitwise_and(inputFrame, filterMask2, output);

	//output = rawFg;

	auto stop = std::chrono::high_resolution_clock::now();

	this->microseconds = std::chrono::duration_cast<std::chrono::microseconds>(stop - start).count();
}

void Detector::processFrameGPU(const cv::UMat& inputFrame)
{
	//auto start = std::chrono::high_resolution_clock::now();

	//cv::cuda::GpuMat input;

	//input.upload(inputFrame);

	//cv::cuda::GpuMat rawFg;
	//bgSubtractor->apply(input, rawFg);

	//cv::cuda::GpuMat threshold;
	//cv::cuda::threshold(rawFg, threshold, 127, 255, cv::ThresholdTypes::THRESH_BINARY);


	//cv::Ptr<cv::cuda::Filter> median = cv::cuda::createMedianFilter(CV_8UC1, 5);
	//cv::cuda::GpuMat blur;
	//median->apply(threshold, blur);

	//auto structEleme = cv::getStructuringElement(cv::MorphShapes::MORPH_RECT, cv::Size(5, 5));
	//auto structEleme2 = cv::getStructuringElement(cv::MorphShapes::MORPH_RECT, cv::Size(9, 9));

	//cv::cuda::GpuMat morph1;
	//cv::cuda::GpuMat morph2;
	//cv::Ptr<cv::cuda::Filter> morphEx1 = cv::cuda::createMorphologyFilter(cv::MorphTypes::MORPH_CLOSE, CV_8UC1, structEleme);
	//morphEx1->apply(blur, morph1);
	//cv::Ptr<cv::cuda::Filter> morphEx2 = cv::cuda::createMorphologyFilter(cv::MorphTypes::MORPH_OPEN, CV_8UC1, structEleme2);
	//morphEx1->apply(morph1, morph2);

	//cv::UMat morph2Finished;
	//morph2.download(morph2Finished);

	//std::vector<std::vector<cv::Point> > contours;
	//std::vector<cv::Vec4i> hierarchy;
	//cv::findContours(morph2Finished, contours, hierarchy, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);


	//cv::UMat filterMask = cv::UMat::zeros(inputFrame.size(), CV_8UC1);
	//for (int i = 0; i < contours.size(); i++)
	//{
	//	double area = cv::contourArea(contours[i]);
	//	if (area > 20) {
	//		cv::drawContours(filterMask, contours, i, cv::Scalar(255), -1);
	//	}
	//}

	//cv::UMat fiiltered;

	//auto structEleme3 = cv::getStructuringElement(cv::MorphShapes::MORPH_ELLIPSE, cv::Size(29, 29));
	//cv::morphologyEx(filterMask, fiiltered, cv::MorphTypes::MORPH_DILATE, structEleme3);

	//cv::UMat filterMask2 = cv::UMat::zeros(inputFrame.size(), CV_8UC3);
	//cv::findContours(fiiltered, contours, hierarchy, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

	//for (int i = 0; i < contours.size(); i++)
	//{
	//	double area = cv::contourArea(contours[i]);
	//	if (area > 20)
	//	{
	//		cv::rectangle(filterMask2, cv::boundingRect(contours[i]), cv::Scalar(255, 255, 255), -1);
	//	}
	//}

	//cv::bitwise_and(inputFrame, filterMask2, output);

	////output = rawFg;

	//auto stop = std::chrono::high_resolution_clock::now();

	//this->microseconds = std::chrono::duration_cast<std::chrono::microseconds>(stop - start).count();
}

cv::UMat& Detector::getOutput()
{
	return output;
}

File: python/detector.py
import cv2
import numpy as np


class Detector:

    def __init__(self):
        self.fgbg = cv2.createBackgroundSubtractorMOG2(150)
        self.openKernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (9, 9))
        self.closeKernel = cv2.getStructuringElement(
            cv2.MORPH_ELLIPSE, (21, 21))

    def __detect(self, frame):
        self.rawFgmask = self.fgbg.apply(frame)

        _, fgmask = cv2.threshold(self.rawFgmask, 127, 255, cv2.THRESH_BINARY)
        fgmask = cv2.medianBlur(fgmask, 7)
        fgmask = cv2.morphologyEx(fgmask, cv2.MORPH_OPEN, self.openKernel)
        fgmask = cv2.morphologyEx(fgmask, cv2.MORPH_CLOSE, self.closeKernel)
        # fgmask = cv2.morphologyEx(fgmask, cv2.MORPH_DILATE, openKernel,iterations=4)
        # fgmask = cv2.morphologyEx(fgmask, cv2.MORPH_CLOSE, kernel, iterations=1)
        self.fgmask = fgmask

    def __getRectMask(self, fgmask):
        contours, hierarchy = cv2.findContours(
            fgmask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

        fgmaskFilter = fgmask
        for c in contours:
            area = cv2.contourArea(c)

            if area < 20:
                cv2.fillPoly(fgmaskFilter, pts=[c], color=0)
                continue

        fgmaskFilter = cv2.morphologyEx(
            fgmaskFilter, cv2.MORPH_CLOSE, cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (51, 51)))

        contours, hierarchy = cv2.findContours(
            fgmaskFilter, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        contoursImg = np.empty(fgmask.shape, np.uint8)

        cv2.drawContours(contoursImg, contours, -1, 255, 3)

        rectMask = np.zeros((fgmask.shape[0], fgmask.shape[1], 3), np.uint8)
        contours_poly = [None]*len(contours)
        for i, c in enumerate(contours):
            contours_poly[i] = cv2.approxPolyDP(c, 3, True)
            rect = cv2.boundingRect(contours_poly[i])
            cv2.rectangle(
                rectMask, (rect[0], rect[1]), (rect[0] + rect[2], rect[1] + rect[3]),(255, 255, 255),cv2.FILLED)

        self.rectMask = rectMask
        self.contoursImg = contoursImg

    def processFrame(self, frame):
        self.__detect(frame)
        self.__getRectMask(self.fgmask)
        return self.fgmask, self.rectMask

    def processFrameDebug(self, frame):
        self.__detect(frame)
        self.__getRectMask(self.fgmask)
        return self.rawFgmask, self.fgmask, self.rectMask, self.contoursImg
